src/Workspaces/Core/MSBuild/MSBuild/MSBuildProjectLoader.Worker_ResolveReferences.cs

```diff
@@ -0,0 +1,295 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.MSBuild
+{
+    public partial class MSBuildProjectLoader
+    {
+        private partial class Worker
+        {
+            private struct ResolvedReferences
+            {
+                public ImmutableArray<ProjectReference> ProjectReferences { get; }
+                public ImmutableArray<MetadataReference> MetadataReferences { get; }
+
+                public ResolvedReferences(ImmutableArray<ProjectReference> projectReferences, ImmutableArray<MetadataReference> metadataReferences)
+                {
+                    ProjectReferences = projectReferences;
+                    MetadataReferences = metadataReferences;
+                }
+            }
+
+            private class MetadataReferenceSet
+            {
+                private readonly ImmutableArray<MetadataReference> _metadataReferences;
+                private readonly ImmutableDictionary<string, HashSet<int>> _pathToIndexMap;
+                private readonly HashSet<int> _indecesToRemove;
+
+                public MetadataReferenceSet(ImmutableArray<MetadataReference> metadataReferences)
+                {
+                    _metadataReferences = metadataReferences;
+                    _pathToIndexMap = CreatePathToIndexMap(metadataReferences);
+                    _indecesToRemove = new HashSet<int>();
+                }
+
+                private static ImmutableDictionary<string, HashSet<int>> CreatePathToIndexMap(IEnumerable<MetadataReference> metadataReferences)
+                {
+                    var pathToIndexMap = ImmutableDictionary.CreateBuilder<string, HashSet<int>>(PathUtilities.Comparer);
+
+                    var index = 0;
+                    foreach (var metadataReference in metadataReferences)
+                    {
+                        var filePath = GetFilePath(metadataReference);
+                        if (!pathToIndexMap.TryGetValue(filePath, out var indeces))
+                        {
+                            indeces = new HashSet<int>();
```

Comment:
not sure how this needs to be set?

Reply:
will there ever be duplicates? the code looks like it wont?

Reply:
Yeah, there can be duplicates when extern aliases are used. We have a test where a metadata reference has _two_ extern aliases. In that case, there are two MetadataReferences passed to the compiler with the same path but different aliases. `MSBuildWorkspaceTests.Test_SharedMetadataReferencesWithAliases` tests this scenario. In that test, the `/reference` command line arguments passed to the compiler look like so:

* `/reference:"C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5\mscorlib.dll"`
* `/reference:"C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5\System.Core.dll"`
* `/reference:Sys1="C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5\System.dll"`
* `/reference:Sys2="C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5\System.dll"`
* `/reference:"C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework\.NETFramework\v4.5\System.Xml.dll"`